### PR TITLE
Add map_index to TaskFail

### PR DESCRIPTION
**Pull request by dstandish**

### CODE REVIEW

File: airflow/migrations/versions/0105_48925b2719cb_add_map_index_to_taskfail.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+def upgrade():
+    """Apply Add map_index to TaskFail"""
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('task_fail', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('run_id', sa.String(length=250), nullable=False))
+        batch_op.add_column(sa.Column('map_index', sa.Integer(), server_default='-1', nullable=False))
+        batch_op.drop_index('idx_task_fail_dag_task_date')
+        batch_op.create_index('idx_task_fail_dag_task_date', ['dag_id', 'task_id', 'run_id'], unique=False)
+        batch_op.create_foreign_key(
+            'sla_miss_ti_fkey',
```
Comment (ashb) [inline airflow/migrations/versions/0105_48925b2719cb_add_map_index_to_taskfail.py]: Wrong table name here :)

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1800,7 +1800,15 @@ def handle_failure(
 
             # Log failure duration
             dag_run = self.get_dagrun(session=session)  # self.dag_run not populated by refresh_from_db
-            session.add(TaskFail(task, dag_run.execution_date, self.start_date, self.end_date))
+            session.add(
+                TaskFail(
+                    task=task,
+                    run_id=dag_run.run_id,
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
                    run_id=self.run_id,
```

**Review by ashb — COMMENTED** [review]
Few typos, change looks okay

File: airflow/migrations/versions/0105_48925b2719cb_add_map_index_to_taskfail.py (hunk 1 above)
Comment (dstandish) [reply airflow/migrations/versions/0105_48925b2719cb_add_map_index_to_taskfail.py]: oops!

File: airflow/models/taskfail.py [hunk 3]
```diff
@@ -16,34 +16,59 @@
...
+from typing import TYPE_CHECKING
 
-from airflow.models.base import COLLATION_ARGS, ID_LEN, Base
+from sqlalchemy import Column, ForeignKeyConstraint, Integer
+
+from airflow.models.base import Base, StringID
 from airflow.utils.sqlalchemy import UtcDateTime
 
+if TYPE_CHECKING:
+    pass
```
Comment (uranusjr) [inline airflow/models/taskfail.py]: Stray code?

File: airflow/www/views.py [hunk 4]
```diff
@@ -2847,8 +2847,8 @@ def duration(self, dag_id, session=None):
             session.query(TaskFail)
             .filter(
                 TaskFail.dag_id == dag.dag_id,
-                TaskFail.execution_date >= min_date,
-                TaskFail.execution_date <= base_date,
+                TaskInstance.execution_date >= min_date,
+                TaskInstance.execution_date <= base_date,
```
Comment (uranusjr) [inline airflow/www/views.py]: I wonder if we could directly use `DagRun.execution_date` here.

Comment (dstandish) [reply airflow/www/views.py]: could try it

File: airflow/models/taskfail.py (hunk 3 above)
Comment (dstandish) [reply airflow/models/taskfail.py]: yeah when the import under IF is no longer used our precommit hook removes the import but leaves the if

File: airflow/models/taskfail.py [hunk 5]
```diff
@@ -28,22 +29,42 @@ class TaskFail(Base):
     __tablename__ = "task_fail"
 
     id = Column(Integer, primary_key=True)
-    task_id = Column(String(ID_LEN, **COLLATION_ARGS), nullable=False)
-    dag_id = Column(String(ID_LEN, **COLLATION_ARGS), nullable=False)
-    execution_date = Column(UtcDateTime, nullable=False)
+    task_id = Column(StringID(), nullable=False)
+    dag_id = Column(StringID(), nullable=False)
+    run_id = Column(StringID(), nullable=False)
+    map_index = Column(Integer, server_default='-1', nullable=False)
```
Comment (dstandish) [inline airflow/models/taskfail.py:35]: question @uranusjr  @ashb ...

the uniqueness of these 4 attributes will be enforced through the foreign key.  do you think that we should also add a unique index though for performance reasons?

File: airflow/www/views.py (hunk 4 above)
Comment (dstandish) [reply airflow/www/views.py]: appears it does work.  that's pretty magical!

**Comment by dstandish** [comment]
@ashb @uranusjr added a helper util for checking a table for dupes and used it for task fail ([here](https://github.com/apache/airflow/pull/22260/commits/7701c099e006e0f74bca0b29a438a836b6b4979a))

File: airflow/models/taskfail.py [hunk 6]
```diff
@@ -28,22 +29,42 @@ class TaskFail(Base):
...
+                "task_instance.run_id",
+                "task_instance.map_index",
+            ],
+            name='task_fail_ti_fkey',
+            ondelete="CASCADE",
+        ),
+    )
 
-    def __init__(self, task, execution_date, start_date, end_date):
+    def __init__(self, task, run_id, start_date, end_date, map_index=None):
```
Comment (uranusjr) [inline airflow/models/taskfail.py]: Should map_inddex default to -1 instead? Is a default needed here at all?

File: tests/core/test_core.py [hunk 7]
```diff
@@ -89,22 +89,25 @@ def test_task_fail_duration(self, dag_maker):
...
             .all()
         )
         op2_fails = (
             session.query(TaskFail)
-            .filter_by(task_id='fail_sleepy', dag_id=dag.dag_id, execution_date=DEFAULT_DATE)
+            .filter(
+                TaskFail.task_id == 'fail_sleepy',
+                TaskFail.dag_id == dag.dag_id,
+                TaskInstance.execution_date == DEFAULT_DATE,
+            )
```
Comment (uranusjr) [inline tests/core/test_core.py]:
```suggestion
            .filter(
                TaskFail.task_id == 'fail_sleepy',
                TaskFail.dag_id == dag.dag_id,
                DagRun.execution_date == DEFAULT_DATE,
            )
```

Same for other calls above.

File: airflow/models/taskfail.py (hunk 6 above)
Comment (dstandish) [reply airflow/models/taskfail.py]: > Is a default needed here at all?

yeah was thinking that better to let it come from task instance.

Comment (dstandish) [reply airflow/models/taskfail.py]: updated

**Review by uranusjr — APPROVED** [review]

**Review by ashb — APPROVED** [review]

**Pull request merged**